### DISCUSSION TITLE
Nullable inheritance for classes.

### DISCUSSION THREAD

**Discussion by RomanSoloweow** [dotnet/csharplang#5410] [Language Ideas]
How to implement a generic method that would return null for classes and structures as well?

**Comment by Joe4evr** [comment]
You can't. There is no instruction that can take an open `T` and turn it into `null` when it is a reference type and `Nullable<T>` when it is a value type.

**Reply by RomanSoloweow** [reply]
Therefore, I believe that it is necessary to add the ability to make Nullable for classes

**Reply by HaloFour** [reply]
Classes are already nullable, adding a wrapper around it would be redundant.  At this point changing the meaning of `T?` would also be a breaking change.

**Reply by jnm2** [reply]
You can work around this runtime limitation to some extent by declaring overloads that contain `T?` somewhere in their signature and one has `where T : class` and the other has `where T : struct`. E.g.: https://github.com/shouldly/shouldly/blob/037a2b9c118aa266a84988d9adf0301c4ca1ad5a/src/Shouldly/ShouldlyExtensionMethods/ShouldBeNullExtensions.cs#L28-L40

Unfortunately, you can't call such a method from a method that uses an unconstrained `T`.

**Reply by RikkiGibson** [reply]
There has been a bit of exploration of "runtime constraint checking", i.e. allowing you to do something (*extremely* strawman) like:

```cs
M<int>(); // ok
M<string>(); // runtime error

void M<T>()
{
    if (default(T) != null)
    {
        MValueType<T>(); // ok
    }
}

[RuntimeConstraintChecks]
void MValueType<T>() where T : struct { }
```

I don't think this would include any flow analysis of type parameters. That is, the compiler wouldn't do anything to try and help you use the type parameter correctly here.

IIRC, this was meant to make some static abstracts use cases more pleasant, but I don't recall more specifically than that. @davidwrighton may wish to weigh in.
